sklearn/model_selection/_split.py

```diff
@@ -732,6 +735,177 @@ def split(self, X, y, groups=None):
+        # Changelist:
+        # - Refactored function to a class following scikit-learn KFold
+        #   interface.
+        # - Added heuristic for assigning group to the least populated fold in
+        #   cases when all other criteria are equal
+        # - Swtch from using python ``Counter`` to ``np.unique`` to get class
+        #   distribution
+        # - Added scikit-learn checks for input: checking that target is binary
+        #   or multiclass, checking passed random state, checking that number
+        #   of splits is less than number of members in each class, checking
+        #   that least populated class has more members than there are splits.
+        rng = check_random_state(self.random_state)
+        y = np.asarray(y)
+        type_of_target_y = type_of_target(y)
+        allowed_target_types = ('binary', 'multiclass')
+        if type_of_target_y not in allowed_target_types:
+            raise ValueError(
+                'Supported target types are: {}. Got {!r} instead.'.format(
+                    allowed_target_types, type_of_target_y))
+
+        y = column_or_1d(y)
+        _, y_inv, y_cnt = np.unique(y, return_inverse=True, return_counts=True)
+        if np.all(self.n_splits > y_cnt):
+            raise ValueError("n_splits=%d cannot be greater than the"
+                             " number of members in each class."
+                             % (self.n_splits))
+        n_smallest_class = np.min(y_cnt)
+        if self.n_splits > n_smallest_class:
+            warnings.warn(("The least populated class in y has only %d"
+                           " members, which is less than n_splits=%d."
+                           % (n_smallest_class, self.n_splits)), UserWarning)
+        labels_num = len(y_cnt)
+        y_counts_per_group = defaultdict(lambda: np.zeros(labels_num))
+        for label, group in zip(y_inv, groups):
+            y_counts_per_group[group][label] += 1
+
+        y_counts_per_fold = defaultdict(lambda: np.zeros(labels_num))
+        groups_per_fold = defaultdict(set)
+
+        groups_and_y_counts = list(y_counts_per_group.items())
+        if self.shuffle:
+            rng.shuffle(groups_and_y_counts)
+
+        for group, y_counts in sorted(groups_and_y_counts,
+                                      key=lambda x: -np.std(x[1])):
+            best_fold = None
+            min_eval = np.inf
+            min_samples_in_fold = np.inf
+            for i in range(self.n_splits):
+                y_counts_per_fold[i] += y_counts
+                std_per_label = []
+                for label in range(labels_num):
+                    std_per_label.append(np.std(
+                        [y_counts_per_fold[j][label] / y_cnt[label]
+                         for j in range(self.n_splits)]))
+                y_counts_per_fold[i] -= y_counts
+                fold_eval = np.mean(std_per_label)
+                samples_in_fold = np.sum(y_counts_per_fold[i])
```

Comment:
Just to check: is this intentionally after removing `y_counts` from `y_counts_per_fold`?

Reply:
Yes, the idea is to find the fold with least number of samples before the addition of the current group (which happens several lines earlier)